### PR TITLE
SPARQL-star: Add embedded triple pattern to PrimaryExpression, define evaluation.

### PR DESCRIPTION
**Pull request by afs**
Addresses issue #163.

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/rdf-star/pull/167.html" title="Last updated on May 6, 2021, 8:10 AM UTC (950082c)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/rdf-star/167/ce2db5e...950082c.html" title="Last updated on May 6, 2021, 8:10 AM UTC (950082c)">Diff</a>

### CODE REVIEW

**Comment by Tpt** [comment]
I am not sure it is possible to reuse the `EmbTP ` rule here without restrictions: `EmbTP ` allows blank nodes that are not allowed in SPARQL expressions. What about introducing a new rule:

```
EmbTPExpr ::= '<<' PrimaryExpression PrimaryExpression PrimaryExpression '>>'
```

and make the produced value an alternative writing of the `TRIPLE` function?

**Comment by afs** [comment]
Thanks for pointing that out.

PR set to "draft".

The idea is that it should be like a term in a graph pattern, so it is not full `PrimaryExpression` which allows `( expression )` and functions and built-ins. 

I think it is going to have to be written out as a specific form of vars or constants, and also `a` for the predicate slot. (And so avoids going near related matching forms SPARQL 1.1 does not have `?str@en` and `?lex^^:datatype`.)

For evaluation definition, `TRIPLE` can be used as it covers all the cases (and more).

**Comment by afs** [comment]
Pushed an update for the grammar.

`ExprVarOrTerm` can be written using 177/DataValueTerm but that is currently specifically for VALUES or using `iriOrVar` but as rule names have implications so it seemed clearer to enumerate the cases.

It will need some stylistic reorganisation to fit with the grammar section.  Rules119a and 119b are placed close to use for clarity for now but style is to group new productions at the end.

File: cg-spec/editors_draft.html [hunk 1]
```diff
@@ -895,6 +895,47 @@ <h2>Grammar</h2>
...
+            <a data-cite="SPARQL11-QUERY#rBooleanLiteral">BooleanLiteral</a> `|`
+            <a data-cite="SPARQL11-QUERY#rVar">Var</a>
+            <div style="font-weight: bold">
+            `|` <a href="#rExprEmbTP" style="font-weight: bold">ExprEmbTP</a>
+            </div>
+          </td>
+        </tr>
+        <tr id="rExprEmbTP">
+          <td>[119a]</td>
+          <td style="font-weight: bold">`ExprEmpTP`</td>
```
Comment (gkellogg) [inline cg-spec/editors_draft.html:917]:
```suggestion
          <td style="font-weight: bold">`ExprEmbTP`</td>
```

File: cg-spec/editors_draft.html [hunk 2]
```diff
@@ -895,6 +895,47 @@ <h2>Grammar</h2>
...
+          <td>[119b]</td>
+          <td style="font-weight: bold">`ExprVarOrTerm`</td>
+          <td style="font-weight: bold">::=</td>
+          <td style="font-weight: bold">
+            <a data-cite="SPARQL11-QUERY#riri">iri</a> `|` 
+            <a data-cite="SPARQL11-QUERY#rRDFLiteral">RDFLiteral</a> `|` 
+            <a data-cite="SPARQL11-QUERY#rNumericLiteral">NumericLiteral</a> `|`
+            <a data-cite="SPARQL11-QUERY#rBooleanLiteral">BooleanLiteral</a> `|`
+            <a data-cite="SPARQL11-QUERY#rVar">Var</a>
+          </td>
```
Comment (gkellogg) [inline cg-spec/editors_draft.html:937]: Should this include `ExprVarOrTerm` in the same way that `EmbSubjectOrObject` includes `EmbTP`? Or, is that not possible in this context?

**Review by gkellogg — CHANGES_REQUESTED** [review]
See my inline suggestions and questions.

Without adding `ExprEmbTP` to `ExprVarOrTerm` I can't parse one of our syntax tests.

File: cg-spec/editors_draft.html (hunk 1 above)
Comment (afs) [reply cg-spec/editors_draft.html]: Thanks! Done.

File: cg-spec/editors_draft.html (hunk 2 above)
Comment (afs) [reply cg-spec/editors_draft.html:938]: Done.

**Comment by afs** [comment]
This PR now has:

* Grammar rules for `PrimaryExpression` involving `<<>>`.
* Two new bad syntax tests for `<<>> in expressions (the "good" case was already covered).
* Definition of evaluation of `<<>>` in an expression context (added new section 4.4.6). It is defined as being the same as `TRIPLE` with the expressions allowed by the syntax for `<<>>`.

**Review by gkellogg — APPROVED** [review]
This looks good; I have yet to try running your new tests, though.

**Comment by afs** [comment]
@pchampin This now adds a section 4.4.6 for evaluation of `<< >>` in expressions.

**Review by pchampin — APPROVED** [review]

**Pull request merged**